src/liballoc/vec.rs

```diff
@@ -1984,17 +1986,31 @@ impl<'a, T> IntoIterator for &'a mut Vec<T> {
 impl<T> Extend<T> for Vec<T> {
     #[inline]
     fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
-        <Self as SpecExtend<T, I::IntoIter>>::spec_extend(self, iter.into_iter())
+        if self.capacity() > 0 {
+            <Self as SpecExtend<T, I::IntoIter>>::spec_extend(self, iter.into_iter())
+        } else {
+            // if self has no allocation then use the more powerful from_iter specializations
+            let other = SpecFrom::from_iter(iter.into_iter());
+            // replace self, don't run drop since self was empty
+            unsafe {
```

Comment:
How much does using unsafe to avoid the drop impl matter here? It seems like it should in principle be optimized away anyway since the capacity check has already been performed.

Reply:
I had issues `mem::replace` being measurably worse in the in-place specialization and fixed that by manually overwriting the `IntoIter`. So I had "replace bad, overwrite instead" in my head and applied that here too.

I don't recall measuring it in this place. I'll give it a try if you want.

Reply:
I'd prefer to avoid unsafety if it's not necessary.

Reply:
I added a benchmark and couldn't see any difference, so I have replaced the unsafe code.